llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp

```diff
@@ -1886,7 +1899,8 @@ bool AMDGPUDAGToDAGISel::SelectScratchSVAddr(SDNode *N, SDValue Addr,
     return false;
   }
 
-  if (!isFlatScratchBaseLegal(SAddr) || !isFlatScratchBaseLegal(VAddr))
+  if (!isFlatScratchBaseLegal(Addr, SAddr) ||
```

Comment:
With a second look, I find the code here is wrong now because the newly added check only work when there are **two** offsets. For `SAddr + VAddr + immOffset`, the simple and correct way could be just check that `CurDAG->SignBitIsZero(SAddr) && CurDAG->SignBitIsZero(VAddr)`. I would change like that if no objection.

Reply:
Does this even apply in the scratch case? I thought the signed offsets all worked correctly now?

Reply:
Based on Chapter 11 in: https://www.amd.com/content/dam/amd/en/documents/radeon-tech-docs/instruction-set-architectures/rdna3-shader-instruction-set-architecture-feb-2023_0.pdf. See the instruction field part, both the `ADDR` and `SADDR` are still saying the offset is unsigned.

Reply:
@rampitec I thought this was sorted out already, is the documentation correct? When scratch instructions were initially introduced, they had signed offsets that didn't behave correctly so we treated them as unsigned with a lost bit. I thought we only switched to using them on targets where the signed behavior was corrected. Is this not the case?

Reply:
@petar-avramovic should probably know about some details as the logic was introduced by him.

Reply:
Offset values in register (ADDR and SADDR) are treated as **unsigned** by hardware (I tested this and confirmed incorrect behavior when offset value in register was negative). Imm offset value is **signed**. Documentation looks correct.